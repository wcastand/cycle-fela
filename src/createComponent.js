import { h } from 'snabbdom'
import fela from 'fela'

const { combineRules } = fela

export default function createComponent(rules, selector = 'div') {
  return function(props, children) {
    const c = arguments.length === 1 ? props : children
    const p = arguments.length === 1
      ? { component: rules }
      : Object.assign({}, props, { component: rules })
    if (typeof selector === 'string') {
      return h(selector, p, c)
    }
    const sel = selector(p, c)
    const mergedRules = combineRules(sel.data.component, rules)
    const pp = Object.assign({}, p, { component: mergedRules })
    return h(sel.sel, pp, c)
  }
}
