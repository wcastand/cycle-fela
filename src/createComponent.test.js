import createComponent from './createComponent'

test('createComponent should return a function', () =>
  expect(typeof createComponent(() => ({}))).toBe('function'))

test('createComponent second call should return an object', () => {
  const Div = createComponent(() => ({ color: 'red' }))
  const vdom = Div('test')
  const style = vdom.data.component({})

  expect(typeof vdom).toBe('object')
  expect(style).toEqual({ color: 'red' })

  expect(vdom).toMatchSnapshot()
})
test('createComponent should compose', () => {
  const Div = createComponent(() => ({ backgroundColor: 'red' }))
  const RedBoldDiv = createComponent(() => ({ fontWeight: 'bold' }), Div)
  const vdom = RedBoldDiv('test')
  const style = vdom.data.component({})

  expect(vdom).toMatchSnapshot()
  expect(style).toEqual({ backgroundColor: 'red', fontWeight: 'bold' })
})
test('createComponent should compose and overwrite rule', () => {
  const RedDiv = createComponent(() => ({ backgroundColor: 'red' }))
  const YellowDiv = createComponent(
    () => ({ fontWeight: 'bold', backgroundColor: 'yellow' }),
    RedDiv,
  )
  const vdom = YellowDiv('test')
  const style = vdom.data.component({})

  expect(vdom).toMatchSnapshot()
  expect(style).toEqual({ backgroundColor: 'yellow', fontWeight: 'bold' })
})

test('createComponent should have a function in component props', () => {
  const Div = createComponent(() => ({ backgroundColor: 'red' }))
  const vdom = Div('test')

  expect(vdom.sel).toBe('div')
  expect(typeof vdom.data.component).toBe('function')
})
test('createComponent should have a string as text', () => {
  const Div = createComponent(() => ({ backgroundColor: 'red' }))
  const vdom = Div('test')

  expect(vdom.text).toBe('test')
})
test('createComponent should have an array as children', () => {
  const Div = createComponent(() => ({ backgroundColor: 'red' }))
  const vdom = Div([Div('test'), Div('test2')])
  expect(vdom.children.length).toBe(2)
})
