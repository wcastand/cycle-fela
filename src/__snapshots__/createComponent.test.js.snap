// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createComponent second call should return an object 1`] = `
Object {
  "children": undefined,
  "data": Object {
    "component": [Function],
  },
  "elm": undefined,
  "key": undefined,
  "sel": "div",
  "text": "test",
}
`;

exports[`createComponent should compose 1`] = `
Object {
  "children": undefined,
  "data": Object {
    "component": [Function],
  },
  "elm": undefined,
  "key": undefined,
  "sel": "div",
  "text": "test",
}
`;

exports[`createComponent should compose and overwrite rule 1`] = `
Object {
  "children": undefined,
  "data": Object {
    "component": [Function],
  },
  "elm": undefined,
  "key": undefined,
  "sel": "div",
  "text": "test",
}
`;
